nengo/cache.py

```diff
@@ -321,6 +321,11 @@ def shrink(self, limit=None):
             Maximum size of the cache in bytes.
         """
         if self.readonly:
+            warnings.warn("Cannot shrink a readonly cache.")
+            return
+
+        if self._index is None:
```

Comment:
Right, we should probably reset `self._index` to `None`. It feels like things should only be able to happen inside the context manager; if it's just that kind of one-time thing then it should be in the `__init__` (but yes, I know `__init__` and `__del__` are not feasible for the cache).